nodeup/pkg/model/kube_apiserver_healthcheck.go

```diff
@@ -0,0 +1,179 @@
+	manifest := b.findHealthcheckManifest()
+	if manifest == nil {
+		return nil
+	}
+
+	p := b.ConfigBase.Join(manifest.Path)
+
+	data, err := p.ReadFile()
+	if err != nil {
+		return fmt.Errorf("error reading kube-apiserver-healthcheck manifest %s: %v", manifest.Path, err)
+	}
+
+	sidecar := &corev1.Pod{}
+	if err := yaml.Unmarshal(data, sidecar); err != nil {
+		return fmt.Errorf("error parsing kube-apiserver-healthcheck manifest %s: %v", manifest.Path, err)
+	}
+
+	// Quick-and-dirty merge of the fields we care about
+	pod.Spec.Containers = append(pod.Spec.Containers, sidecar.Spec.Containers...)
+	pod.Spec.Volumes = append(pod.Spec.Volumes, sidecar.Spec.Volumes...)
+
+	return nil
+}
+
+func (b *KubeAPIServerBuilder) addHealthcheckSidecarTasks(c *fi.ModelBuilderContext) error {
+	id := "kube-apiserver-healthcheck"
+	secretsDir := "/etc/kubernetes/" + id + "/secrets"
+	userID := wellknownusers.KubeApiserverHealthcheckID
+	userName := wellknownusers.KubeApiserverHealthcheckName
+
+	// We create user a user and hardcode its UID to 10012 as
+	// that is the ID used inside the container.
+	{
+		c.AddTask(&nodetasks.UserTask{
+			Name:  userName,
+			UID:   userID,
+			Shell: "/sbin/nologin",
+			Home:  secretsDir,
+		})
+	}
+
+	clientKey, clientCert, err := b.buildClientKeypair(id)
+	if err != nil {
+		return err
+	}
+
+	c.AddTask(&nodetasks.File{
+		Path: filepath.Join(secretsDir),
+		Type: nodetasks.FileType_Directory,
+		Mode: s("0755"),
+	})
+
+	c.AddTask(&nodetasks.File{
+		Path:     filepath.Join(secretsDir, "client.crt"),
+		Contents: fi.NewBytesResource(EncodeCertPEM(clientCert)),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0644"),
+		Owner:    s(userName),
+	})
+	c.AddTask(&nodetasks.File{
+		Path:     filepath.Join(secretsDir, "client.key"),
+		Contents: fi.NewBytesResource(EncodePrivateKeyPEM(clientKey)),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0600"),
+		Owner:    s(userName),
+	})
+
+	cert, err := b.GetCert(fi.CertificateId_CA)
+	if err != nil {
+		return err
+	}
+
+	c.AddTask(&nodetasks.File{
+		Path:     filepath.Join(secretsDir, "ca.crt"),
+		Contents: fi.NewBytesResource(cert),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0644"),
+		Owner:    s(userName),
+	})
+
+	return nil
+}
+
+func (b *KubeAPIServerBuilder) buildClientKeypair(id string) (*rsa.PrivateKey, *x509.Certificate, error) {
+	signerID := fi.CertificateId_CA
+
+	var signerKey *pki.PrivateKey
+	{
+		k, err := b.KeyStore.FindPrivateKey(signerID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if k == nil {
+			return nil, nil, fmt.Errorf("private key %q not found", signerID)
+		}
+		signerKey = k
+	}
+
+	var signerCertificate *pki.Certificate
+	{
+		cert, err := b.KeyStore.FindCert(signerID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if cert == nil {
+			return nil, nil, fmt.Errorf("certificate %q not found", signerID)
+		}
+
+		signerCertificate = cert
+	}
+
+	privateKey, err := NewPrivateKey(2048)
```

Comment:
Please use `pki.GeneratePrivateKey()` and don't hardcode the key size.

Reply:
I was thinking about creating simpler methods, but looking at pki, that's exactly what you've already done!